core/lib/spree/core/engine.rb

```diff
@@ -105,6 +105,12 @@ class Engine < ::Rails::Engine
         ]
       end
 
+      initializer 'spree.promo.register.promotions.shipping_actions', before: :load_config_initializers do |app|
```

Comment:
Got it, thanks for taking the time to explain this better. IMO it's reasonable to care about the above scenario. As you said, there could be promotions that would stop working if a developer removes that class from the list before all promos with that actions are expired or disabled.

I'm fine merging and iterating on this, maybe with time we'll be able to simplify this logic a lot since it seems like we just want to automatically try to apply some promotions after a specific checkout step (probably I'm too optimistic and it's much more complex than this 😄 ).